Automattic/wp-calypso
# fix: override "View Post" behavior for wp-desktop

### Description

A bug in the WordPress Desktop app was recently discovered in which clicking `View Post` on a newly updated/published post takes the user out of the Calypso iframe context (since the window location is overriden), leaving the app in an unusable state (all keyboard shortcuts cease to work, and clicking on items in the sidebar panel will take us to a browser, instead of remaining in-app). 

~~I've been trying to override this behavior in this PR, but haven't quite gotten it to work (see comments below). I would appreciate some pointers/suggestions on what I may be missing.~~

**_Update_**: Finally got this working and tested in a wp.com sandbox. ✅ 

Fixes Automattic/wp-desktop#908 (see issue for more details) and pairs with Automattic/wp-desktop#924.

We'd want to override the default behavior wherever "View Post" is available to be clicked, such as the snackbar in the bottom left and under "Publish".

<p align="center">
<img width="250" alt="view-post-publish" src="https://user-images.githubusercontent.com/8979548/84699467-906b0080-af1f-11ea-8c55-f96103bd6199.png">
</p>

<p align="center">
<img width="250" alt="view-post-snackbar" src="https://user-images.githubusercontent.com/8979548/84699469-906b0080-af1f-11ea-864f-2866dc503f17.png">
</p>

### To Reproduce

In the desktop app:

1. Edit an existing post and click "View Post" in the snackbar, OR
2. Publish a draft and click "View Post" when published

_Expected_: Navigate to the published post in an external browser
_Actual_: The published post is displayed by overriding the Electron window's location, and because the iframe context has been lost the user is left in a broken state in-app.

<p align="center">
<img width="307" alt="out-of-calypso-iframe" src="https://user-images.githubusercontent.com/8979548/84700800-d4f79b80-af21-11ea-9170-5e800eaa406d.png">
</p>

### To Test

Using a wp.com sandbox (sandbox the block editor as well as a test site), verify that:

- These changes do not affect the existing behavior of clicking "View Post" after a post is updated or published.
- These changes allow the desktop app as built from Automattic/wp-desktop#924 to open the post's link in an external browser when "View Post" is clicked
